File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        # store any values we see and store the index so that we can have
        # duplicates that dont have the same combination
        seen = {}
        result = []
        candidates.sort()
        
        def dfs(index, curr, total) -> None:
            if total == target:
                result.append(curr.copy())
                return 
            if total > target or index >= len(candidates):
                return 
            curr.append(candidates[index])
            dfs(index + 1, curr, total + candidates[index])
            curr.pop()

            while index + 1 < len(candidates) and candidates[index] == candidates[index + 1]:
                index += 1
            
            dfs(index + 1, curr, total)
            
        
        dfs(0, [], 0)
        return result

File: Data Structures & Algorithms/rotate-array/submission-2.java
class Solution {
    public void rotate(int[] nums, int k) {
        int[] result = new int[nums.length];

        for (int i = 0; i < nums.length; i++){
            int index = (i + k) % nums.length;
            result[index] = nums[i];
        }

        for (int j = 0; j < nums.length; j++){
            nums[j] = result[j];
        }
    }
}